Dedup step 1

gfr10598 opened the pull request:
Some utility functions for dealing with Table objects, and listing, and getting basic and detailed info about Tables.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/etl/392)
<!-- Reviewable:end -->

stephen-soltesz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 6 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[dedup/dedup.go, line 96 at r1](https://reviewable.io:443/reviews/m-lab/etl/392#-L3d-9groeaR0zN_QNZ0:-L3d-9groeaR0zN_QNZ1:bq8gbap) ([raw file](https://github.com/m-lab/etl/blob/9f08d36e9a644a895b3a27527d5244da67b6a50d/dedup/dedup.go#L96)):*
> ```Go
> 		FROM (
> 			-- This avoids null counts when the partition doesn't exist or is empty.
>   		    SELECT 0 AS tests, "fake-task" AS Task 
> ```

Do these settings work for you in vscode? (although it might not within quoted text? idk)
```
  "editor.trimAutoWhitespace": true,
  "files.trimTrailingWhitespace": true,  
```

---

*[dedup/dedup.go, line 105 at r1](https://reviewable.io:443/reviews/m-lab/etl/392#-L3d09IFOCBlylQ5z1q7:-L3d09IFOCBlylQ5z1q8:bswp0o7) ([raw file](https://github.com/m-lab/etl/blob/9f08d36e9a644a895b3a27527d5244da67b6a50d/dedup/dedup.go#L105)):*
> ```Go
> 
> 	// TODO - this should take a context?
> 	err := dsExt.QueryAndParse(queryString, &detail)
> ```

In epoxy client, I'm starting to experiment with context.WithTimeout to make sure network requests or commands terminate eventually. My thought is that in principle adding context timeouts to all network requests is probably good policy. Just to say, I approve of this TODO :)

---

*[dedup/dedup.go, line 140 at r1](https://reviewable.io:443/reviews/m-lab/etl/392#-L3d2RYZ4-WH1MRQpzxd:-L3d2RYZ4-WH1MRQpzxe:b-q46q0j) ([raw file](https://github.com/m-lab/etl/blob/9f08d36e9a644a895b3a27527d5244da67b6a50d/dedup/dedup.go#L140)):*
> ```Go
> // tableNameParts is used to describe a templated table or table partition.
> type tableNameParts struct {
> 	prefix, yyyymmdd string
> ```

What do you think about adopting a convention of keeping each field in a separate line?

---

*[dedup/dedup.go, line 156 at r1](https://reviewable.io:443/reviews/m-lab/etl/392#-L3d3ZMKuvZAiiVxiuag:-L3d3ZMKuvZAiiVxiuah:bx81sx5) ([raw file](https://github.com/m-lab/etl/blob/9f08d36e9a644a895b3a27527d5244da67b6a50d/dedup/dedup.go#L156)):*
> ```Go
> // The project/dataset/table/partition may or may not actually exist.
> // This does NOT do any network operations.
> func getTable(bqClient *bigquery.Client, project, dataset, table, partition string) (*bigquery.Table, error) {
> ```

I don't understand this function. I don't see callers, and the behavior is confusing, e.g. "this should fail".

---

*[dedup/dedup.go, line 173 at r1](https://reviewable.io:443/reviews/m-lab/etl/392#-L3d5axK-dBdScgiN5Ht:-L3d5axK-dBdScgiN5Hu:b-leoawx) ([raw file](https://github.com/m-lab/etl/blob/9f08d36e9a644a895b3a27527d5244da67b6a50d/dedup/dedup.go#L173)):*
> ```Go
> 	tableName := table.TableID
> 	parts := strings.Split(tableName, "$")
> 	if len(parts) != 2 {
> ```

Use `getTableParts` and check whether `parts.isPartition` is true?

---

*[dedup/dedup.go, line 176 at r1](https://reviewable.io:443/reviews/m-lab/etl/392#-L3d5lvG-guLaZLJ77sH:-L3d5lvG-guLaZLJ77sI:bo28j0t) ([raw file](https://github.com/m-lab/etl/blob/9f08d36e9a644a895b3a27527d5244da67b6a50d/dedup/dedup.go#L176)):*
> ```Go
> 		return bqext.PartitionInfo{}, errors.New("TableID missing partition: " + tableName)
> 	}
> 	fullTable := fmt.Sprintf("%s:%s.%s", table.ProjectID, table.DatasetID, parts[0])
> ```

While small, this could be a good candidate for bqext -- is it reasonable for calling code to not need to know how to format a bq table?

---

*[dedup/dedup.go, line 187 at r1](https://reviewable.io:443/reviews/m-lab/etl/392#-L3d6M1ufuJH3TkK2ImG:-L3d6M1ufuJH3TkK2ImH:b-j13sms) ([raw file](https://github.com/m-lab/etl/blob/9f08d36e9a644a895b3a27527d5244da67b6a50d/dedup/dedup.go#L187)):*
> ```Go
> 		FROM
> 		  [%s$__PARTITIONS_SUMMARY__]
> 		WHERE partition_id = "%s" `, fullTable, parts[1])
> ```

`parts.yyyymmdd` would read more easily than `parts[1]`

---

*[dedup/README.md, line 41 at r1](https://reviewable.io:443/reviews/m-lab/etl/392#-L3dQBPDwob9ybpRsAxD:-L3dQBPDwob9ybpRsAxE:b-3855jm) ([raw file](https://github.com/m-lab/etl/blob/9f08d36e9a644a895b3a27527d5244da67b6a50d/dedup/README.md#L41)):*
> ```Markdown
> 
> ## Other considerations:
> 1. Should run the tar file check.  Unfortunately, there may be tar
> ```

What is "the tar file check"?

---

*[etl/globals.go, line 12 at r1](https://reviewable.io:443/reviews/m-lab/etl/392#-L3d7TvJy1KAl_daTpoN:-L3d7TvJy1KAl_daTpoO:b-xyuqzx) ([raw file](https://github.com/m-lab/etl/blob/9f08d36e9a644a895b3a27527d5244da67b6a50d/etl/globals.go#L12)):*
> ```Go
> 
> // YYYYMMDD is a regexp string for identifying dense dates.
> const YYYYMMDD = `\d{4}[01]\d[0123]\d`
> ```

Why the alternate declaration form?

Why delete the MlabDomain comment?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/392#-:-L3dQLl-X5cklRcIhRq6:bnfp4nl)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
Review status: 0 of 6 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[dedup/dedup.go, line 96 at r1](https://reviewable.io:443/reviews/m-lab/etl/392#-L3d-9groeaR0zN_QNZ0:-L3dRXJ_AlDX2qdt3CG7:b-9fbbux) ([raw file](https://github.com/m-lab/etl/blob/9f08d36e9a644a895b3a27527d5244da67b6a50d/dedup/dedup.go#L96)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Do these settings work for you in vscode? (although it might not within quoted text? idk)
```
  "editor.trimAutoWhitespace": true,
  "files.trimTrailingWhitespace": true,  
```

</blockquote></details>

Yes but they got blown away with recent update.  I will add them again.

---

*[dedup/dedup.go, line 105 at r1](https://reviewable.io:443/reviews/m-lab/etl/392#-L3d09IFOCBlylQ5z1q7:-L3dV0RQUrXWCr3lN8ky:bb6sh2i) ([raw file](https://github.com/m-lab/etl/blob/9f08d36e9a644a895b3a27527d5244da67b6a50d/dedup/dedup.go#L105)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

In epoxy client, I'm starting to experiment with context.WithTimeout to make sure network requests or commands terminate eventually. My thought is that in principle adding context timeouts to all network requests is probably good policy. Just to say, I approve of this TODO :)
</blockquote></details>

Ack.

---

*[dedup/dedup.go, line 140 at r1](https://reviewable.io:443/reviews/m-lab/etl/392#-L3d2RYZ4-WH1MRQpzxd:-L3dSB5_TeuBXHgIKWKQ:b-g87hip) ([raw file](https://github.com/m-lab/etl/blob/9f08d36e9a644a895b3a27527d5244da67b6a50d/dedup/dedup.go#L140)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

What do you think about adopting a convention of keeping each field in a separate line?
</blockquote></details>

I would support that.

---

*[dedup/dedup.go, line 156 at r1](https://reviewable.io:443/reviews/m-lab/etl/392#-L3d3ZMKuvZAiiVxiuag:-L3dSkndooP9BF366e2R:begw0lt) ([raw file](https://github.com/m-lab/etl/blob/9f08d36e9a644a895b3a27527d5244da67b6a50d/dedup/dedup.go#L156)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

I don't understand this function. I don't see callers, and the behavior is confusing, e.g. "this should fail".
</blockquote></details>

How about now?

---

*[dedup/dedup.go, line 173 at r1](https://reviewable.io:443/reviews/m-lab/etl/392#-L3d5axK-dBdScgiN5Ht:-L3dUJHxwnNCW6YIKEBq:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/9f08d36e9a644a895b3a27527d5244da67b6a50d/dedup/dedup.go#L173)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Use `getTableParts` and check whether `parts.isPartition` is true?
</blockquote></details>

Done.

---

*[dedup/dedup.go, line 176 at r1](https://reviewable.io:443/reviews/m-lab/etl/392#-L3d5lvG-guLaZLJ77sH:-L3dUPFaAp2RiP2Umq18:bf8ifcn) ([raw file](https://github.com/m-lab/etl/blob/9f08d36e9a644a895b3a27527d5244da67b6a50d/dedup/dedup.go#L176)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

While small, this could be a good candidate for bqext -- is it reasonable for calling code to not need to know how to format a bq table?
</blockquote></details>

If it starts showing up a lot.  Not inclined to worry about it yet.

---

*[dedup/dedup.go, line 187 at r1](https://reviewable.io:443/reviews/m-lab/etl/392#-L3d6M1ufuJH3TkK2ImG:-L3dUVeaqRYe3aI5Xwm7:b-896fix) ([raw file](https://github.com/m-lab/etl/blob/9f08d36e9a644a895b3a27527d5244da67b6a50d/dedup/dedup.go#L187)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

`parts.yyyymmdd` would read more easily than `parts[1]`
</blockquote></details>

Done.

---

*[dedup/README.md, line 41 at r1](https://reviewable.io:443/reviews/m-lab/etl/392#-L3dQBPDwob9ybpRsAxD:-L3dV2OtB6VtN1QIG31c:b-ywciww) ([raw file](https://github.com/m-lab/etl/blob/9f08d36e9a644a895b3a27527d5244da67b6a50d/dedup/README.md#L41)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

What is "the tar file check"?
</blockquote></details>

Fixed comment.

---

*[etl/globals.go, line 12 at r1](https://reviewable.io:443/reviews/m-lab/etl/392#-L3d7TvJy1KAl_daTpoN:-L3dUwzyEqSEFZTe-Ric:b-7uskai) ([raw file](https://github.com/m-lab/etl/blob/9f08d36e9a644a895b3a27527d5244da67b6a50d/etl/globals.go#L12)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

Why the alternate declaration form?

Why delete the MlabDomain comment?
</blockquote></details>

merge accident. fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/392)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[dedup/dedup.go, line 156 at r1](https://reviewable.io:443/reviews/m-lab/etl/392#-L3d3ZMKuvZAiiVxiuag:-L3dYVn3niCDs1L5Y45w:b-837i3v) ([raw file](https://github.com/m-lab/etl/blob/9f08d36e9a644a895b3a27527d5244da67b6a50d/dedup/dedup.go#L156)):*
<details><summary><i>Previously, gfr10598 (Gregory Russell) wrote…</i></summary><blockquote>

How about now?
</blockquote></details>

Okay.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/392#-:-L3dZ6qcqfGYEFdl82GI:bnfp4nl)*
<!-- Sent from Reviewable.io -->
